ooni/ui/web/server.py

```diff
@@ -0,0 +1,608 @@
+        measurements = list_measurements()
+        return self.render_json({"measurements": measurements}, request)
+
+    @app.route('/api/measurement/<string:measurement_id>', methods=["GET"])
+    @xsrf_protect(check=False)
+    @requires_true(attrs=['_is_initialized'])
+    def api_measurement_summary(self, request, measurement_id):
+        try:
+            measurement = get_measurement(measurement_id)
+        except InsecurePath:
+            raise WebUIError(500, "invalid measurement id")
+        except MeasurementNotFound:
+            raise WebUIError(404, "measurement not found")
+        except MeasurementInProgress:
+            raise WebUIError(400, "measurement in progress")
+
+        if measurement['completed'] is False:
+            raise WebUIError(400, "measurement in progress")
+
+        summary = get_summary(measurement_id)
+        return self.render_json(summary, request)
+
+    @app.route('/api/measurement/<string:measurement_id>', methods=["DELETE"])
+    @xsrf_protect(check=True)
+    @requires_true(attrs=['_is_initialized'])
+    def api_measurement_delete(self, request, measurement_id):
+        try:
+            measurement = get_measurement(measurement_id)
+        except InsecurePath:
+            raise WebUIError(500, "invalid measurement id")
+        except MeasurementNotFound:
+            raise WebUIError(404, "measurement not found")
+
+        if measurement['running'] is True:
+            raise WebUIError(400, "Measurement running")
+
+        try:
+            measurement_dir = self.measurement_path.child(measurement_id)
+            measurement_dir.remove()
+        except:
+            raise WebUIError(400, "Failed to delete report")
+
+        return self.render_json({"result": "ok"}, request)
+
+    @app.route('/api/measurement/<string:measurement_id>/keep', methods=["POST"])
+    @xsrf_protect(check=True)
+    @requires_true(attrs=['_is_initialized'])
+    def api_measurement_keep(self, request, measurement_id):
+        try:
+            measurement_dir = self.measurement_path.child(measurement_id)
+        except InsecurePath:
+            raise WebUIError(500, "invalid measurement id")
+
+        summary = measurement_dir.child("keep")
+        with summary.open("w+") as f:
+            pass
+
+        return self.render_json({"status": "ok"}, request)
+
+    @app.route('/api/measurement/<string:measurement_id>/<int:idx>',
+               methods=["GET"])
+    @xsrf_protect(check=False)
+    @requires_true(attrs=['_is_initialized'])
+    def api_measurement_view(self, request, measurement_id, idx):
+        try:
+            measurement_dir = self.measurement_path.child(measurement_id)
+        except InsecurePath:
+            raise WebUIError(500, "Invalid measurement id")
+
+        measurements = measurement_dir.child("measurements.njson")
+
+        # This gets the line idx of the measurement file.
+        # XXX maybe implement some caching here
```

Comment:
A possible way of implementing caching is that you cache the last openned file (or a subset of its lines)